sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+`accessor::get_pointer()`, accessor's `operator []`, C/C++ dereference of an
+USM pointer, local accessors, 2D and 3D accessors.
+Those are temporary restrictions, to be dropped in future.
+
+See auto-generated documentation for the complete list of APIs here. (TBD)
+
+#### USM pointer-based memory access
+##### Flat-address gather/scatter
+perform scattered read/write for the given starting pointer **p** and
+**offsets**.
+
+```cpp
+template <typename T, int n, int ElemsPerAddr = 1,
+	  CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+typename std::enable_if<((n == 16 || n == 32) &&
+    (ElemsPerAddr == 1 || ElemsPerAddr == 2 || ElemsPerAddr == 4)),
+    simd<T, n*ElemsPerAddr>>::type
+flat_load(T *p, simd<uint32_t, n> offsets, simd<uint16_t, n> pred = 1);
+
+template <typename T, int n, int ElemsPerAddr = 1,
+	  CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+typename std::enable_if<((n == 16 || n == 32) &&
+    (ElemsPerAddr == 1 || ElemsPerAddr == 2 || ElemsPerAddr == 4)),
+    void>::type
+flat_store(T *p, simd<T, n*ElemsPerAddr> vals, simd<uint32_t, n> offsets,
+           simd<uint16_t, n> pred = 1);
+```
+##### Flat-address block load/store
+read or write a consecutive block of data for the memory location specified by
+**addr**.
+
+```cpp
+template <typename T, int n, CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+simd<T, n> flat_block_load(const T *const addr);
+
+template <typename T, int n, CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+void flat_block_store(T *p, simd<T, n> vals);
+```
+
+##### Flat-address load4/store4
+perform scattered read/write for the given starting pointer **p** and
+**offsets**. Up to 4 data elements may be accessed at each address depending on
+the enabled channel **Mask**.
+
+```cpp
+template <typename T, int n, ChannelMaskType Mask,
+          CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+    typename std::enable_if<(n == 16 || n == 32),
+                            simd<T, n * __NumChannels(Mask)>>::type
+    flat_load4(T *p, simd<uint32_t, n> offsets, simd<uint16_t, n> pred = 1);
+
+template <typename T, int n, ChannelMaskType Mask,
+          CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+typename std::enable_if<(n == 16 || n == 32), void>::type
+    flat_store4(T *p, simd<T, n * __NumChannels(Mask)> vals,
+            simd<uint32_t, n> offsets, simd<uint16_t, n> pred = 1);
+```
+
+##### Flat-address atomic inc/dec
+perform atomic memory access operation with zero source operand.
+
+```cpp
+template <CmAtomicOpType Op, typename T, int n,
+          CacheHint L1H = CacheHint::Default,
+          CacheHint L3H = CacheHint::Default>
+    typename std::enable_if<__check_atomic<Op, T, n, 0>(), simd<T, n>>::type
+    flat_atomic(T *p, simd<unsigned, n> offset, simd<ushort, n> pred);
```

Comment:
We should look into what can be done to simplify these atomic functions and align their design with the other new atomic features.  That is perhaps out of scope for this PR (the direction is less clear than for the `reduce` changes, for example) but we should open an issue to track it.

Reply:
@Pennycook , can you please open the issue then?

Reply:
I'll open it as soon as this is merged.

Reply:
thanks